packages/rollup-dev-tools/src/tools/transpiler/binary-search-tree.ts

```diff
@@ -0,0 +1,355 @@
+  const searchTreeNodes: BinarySearchNode[] = generateLogSearchTreeNodes(
+    jumpdestIndexesBefore,
+    jumpdestIndexesAfter
+  )
+  log.debug(
+    `successfully generated conceptual log search tree, its flat structure is: \n${JSON.stringify(
+      searchTreeNodes
+    )}`
+  )
+  // The root node is always the final element rerturned by generateLogSearchTreeNodes()
+  const rootNodeId = searchTreeNodes.length - 1
+
+  const bytecode: EVMBytecode = [
+    {
+      opcode: Opcode.JUMPDEST,
+      consumedBytes: undefined,
+    },
+    ...appendAncestorsToBytecode(searchTreeNodes, rootNodeId, true, []), // Recursively fills out the bytecode based on the tree.
+  ]
+  // Fix the PUSH (jumpdests) to be correct once the tree has been generated.
+  const finalBytecode = fixJUMPsToNodes(
+    bytecode,
+    indexOfThisBlock,
+    searchTreeNodes
+  )
+  log.debug(
+    `Generated final bytecode for log search jump table : \n${formatBytecode(
+      finalBytecode
+    )}`
+  )
+  return finalBytecode
+}
+
+/**
+ * Recursively appends to the given bytecode all ancestors for the given NodeId.
+ *
+ * @param treeNodes The binary search tree representing the nodes to process into bytecode
+ * @param nodeId The nodeId to process all ancestors of.
+ * @param isLeftSibling Whether the given nodeId is a left or right sibling (if a right sibling, it will be JUMPed to.)
+ * @param bytecode The existing EVM bytecode to append to.
+ * @returns The EVM Bytecode with all ancestors of the given nodeId having been added.
+ */
+const appendAncestorsToBytecode = (
+  treeNodes: BinarySearchNode[],
+  nodeId: number,
+  isLeftSibling: boolean,
+  bytecode: EVMBytecode
+): EVMBytecode => {
+  const thisNode: BinarySearchNode = treeNodes[nodeId]
+  log.info(
+    `Processing node with Id ${nodeId} of tree into bytecode, its parameters are ${JSON.stringify(
+      thisNode
+    )}.`
+  )
+  if (isLeafNode(thisNode)) {
+    // If this is a leaf node, we can append the leaf and return since no ancestors.
+    bytecode = [
+      ...bytecode,
+      ...generateLeafBytecode(
+        thisNode as BinarySearchLeafNode,
+        nodeId,
+        isLeftSibling
+      ),
+    ]
+  } else {
+    // Otherwise, we process and append the left and right siblings, recursively using this function.
+    // Append this node
+    const thisInternalNode = thisNode as BinarySearchInternalNode
+    bytecode = [
+      ...bytecode,
+      ...generateInternalNodeBytecode(thisInternalNode, nodeId, isLeftSibling),
+    ]
+    // Append left and right children (left first--right siblings are JUMPed to, left siblings not!)
+    const leftChildId = thisInternalNode.leftChildNodeId
+    const rightChildId = thisInternalNode.rightChildNodeId
+    bytecode = appendAncestorsToBytecode(treeNodes, leftChildId, true, bytecode)
+    bytecode = appendAncestorsToBytecode(
+      treeNodes,
+      rightChildId,
+      false,
+      bytecode
+    )
+  }
+  return bytecode
```

Comment:
Mentioned elsewhere, but just to document, if an actual BST, all nodes are the same, simplifying the need to check if leaf node or not and have separate logic. Also allows eager-exit upon traversal instead of requiring traversal to the leaf for every match.